projects/rocprofiler-systems/source/lib/core/trace_cache/perfetto_processor.cpp

```diff
@@ -0,0 +1,1142 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#include "core/trace_cache/perfetto_processor.hpp"
+#include "common.hpp"
+#include "core/agent_manager.hpp"
+#include "core/common_types.hpp"
+#include "core/gpu_metrics.hpp"
+#include "library/tracing.hpp"
+#include "perfetto.hpp"
+#include "trace_cache/metadata_registry.hpp"
+#include "trace_cache/sample_type.hpp"
+#include "trace_cache/storage_parser.hpp"
+
+#include <cstdint>
+#include <nlohmann/json.hpp>
+
+#include <memory>
+#include <mutex>
+#include <string>
+
+#if ROCPROFSYS_USE_ROCM > 0
+#    include "library/rocprofiler-sdk/fwd.hpp"
+#    include <rocprofiler-sdk/context.h>
+#endif
+
+namespace rocprofsys
+{
+namespace trace_cache
+{
+namespace
+{
+struct annotation_entry
+{
+    const char*                                                             key;
+    std::variant<std::string, uint64_t, int64_t, double, int32_t, uint32_t> value;
+};
+
+void
+annotate_perfetto(::perfetto::EventContext&            ctx,
+                  const std::vector<annotation_entry>& annotations)
+{
+    for(const auto& ann : annotations)
+    {
+        std::visit(
+            [&](auto&& val) { tracing::add_perfetto_annotation(ctx, ann.key, val); },
+            ann.value);
+    }
+}
+
+template <typename CategoryT>
+::perfetto::Track
+get_track(CategoryT, std::string name, uint64_t hash_arg)
+{
+    auto  _uuid        = tracing::get_perfetto_category_uuid<CategoryT>(hash_arg);
+    auto& _track_uuids = tracing::get_perfetto_track_uuids();
+
+    if(_track_uuids.find(_uuid) == _track_uuids.end())
+    {
+        const auto _track = ::perfetto::Track(_uuid, ::perfetto::ProcessTrack::Current());
+        auto       _desc  = _track.Serialize();
+
+        _desc.set_name(name);
+        ::perfetto::TrackEvent::SetTrackDescriptor(_track, _desc);
+
+        _track_uuids.emplace(_uuid, name);
+    }
+    return ::perfetto::Track(_uuid, ::perfetto::ProcessTrack::Current());
+}
+
+template <typename Category>
+void
+write_track_data(const struct backtrace_region_sample& _sample, bool use_annotations)
```

Comment:
```suggestion
write_sampling_track_data(const struct backtrace_region_sample& _sample, bool use_annotations)
```

Should naming be fixed here? `write_track_data` is too general